app_test.go

```diff
@@ -113,7 +113,30 @@ func TestNewApp(t *testing.T) {
 			[]string{"Provided", "Provided", "Provided", "Provided", "LoggerInitialized", "Started"},
 			spy.EventTypes())
 
-		assert.Contains(t, spy.Events()[0].(*fxevent.Provided).OutputTypeNames, "struct {}")
+		// Lifecycle, Shutdowner, and DotGraph are indices 0, 1, and 2.
+		// Our type should be index 3.
+		assert.Contains(t, spy.Events()[3].(*fxevent.Provided).OutputTypeNames, "struct {}")
+	})
+
+	t.Run("InternalTypesProvidedFirst", func(t *testing.T) {
+		t.Parallel()
+
+		// Test that internal types get provided before
+		// user-specified types. This increases the chance
+		// that custom loggers relying on internal types
+		// can be built successfully.
+
+		spy := new(fxlog.Spy)
+		app := fxtest.New(t, Provide(func() struct{} { return struct{}{} }),
+			WithLogger(func() fxevent.Logger { return spy }))
+		defer app.RequireStart().RequireStop()
+		require.Equal(t,
+			[]string{"Provided", "Provided", "Provided", "Provided", "LoggerInitialized", "Started"},
+			spy.EventTypes())
+
+		assert.Contains(t, spy.Events()[0].(*fxevent.Provided).OutputTypeNames, "fx.Lifecycle")
```

Comment:
Ah, never mind. You separated it out to its own test.

Consider just putting this in the first test w/o separating it out to its own test.